pkg/broker/handler/pool/pool.go

```diff
@@ -27,25 +30,86 @@ type SyncPool interface {
 	SyncOnce(ctx context.Context) error
 }
 
+type healthChecker struct {
+	mux              sync.RWMutex
+	lastReportTime   time.Time
+	maxStaleDuration time.Duration
+}
+
+func (c *healthChecker) reportHealth() {
+	c.mux.Lock()
+	defer c.mux.Unlock()
+	c.lastReportTime = time.Now()
+}
+
+func (c *healthChecker) lastTime() time.Time {
+	c.mux.RLock()
+	defer c.mux.RUnlock()
+	return c.lastReportTime
+}
+
+func (c *healthChecker) start(ctx context.Context) {
+	c.reportHealth()
+	// TODO: allow changing port? It's pure internal though.
+	srv := &http.Server{
+		Addr:    ":8080",
+		Handler: c,
+	}
+
+	go func() {
+		logging.FromContext(ctx).Info("Starting the sync pool health checker...")
+		if err := srv.ListenAndServe(); err != nil && err != http.ErrServerClosed {
+			logging.FromContext(ctx).Error("the sync pool health checker has stopped unexpectedly", zap.Error(err))
+		}
+	}()
+
+	<-ctx.Done()
+	if err := srv.Shutdown(ctx); err != nil {
+		logging.FromContext(ctx).Error("failed to shutdown the sync pool health checker", zap.Error(err))
+	}
+}
+
+func (c *healthChecker) ServeHTTP(w http.ResponseWriter, req *http.Request) {
+	if req.URL.Path != "/healthz" {
+		w.WriteHeader(http.StatusNotFound)
+		return
+	}
+	// Zero maxStaleDuration means inifinite.
+	if c.maxStaleDuration == 0 {
+		w.WriteHeader(http.StatusOK)
+		return
+	}
+	if time.Now().Sub(c.lastTime()) > c.maxStaleDuration {
+		w.WriteHeader(http.StatusServiceUnavailable)
+		return
+	}
+	w.WriteHeader(http.StatusOK)
+}
+
 // StartSyncPool starts the sync pool.
-func StartSyncPool(ctx context.Context, syncPool SyncPool, syncSignal <-chan struct{}) (SyncPool, error) {
+func StartSyncPool(ctx context.Context, syncPool SyncPool, syncSignal <-chan struct{}, maxStaleDuration time.Duration) (SyncPool, error) {
 	if err := syncPool.SyncOnce(ctx); err != nil {
 		return nil, err
 	}
+	c := &healthChecker{maxStaleDuration: maxStaleDuration}
+	go c.start(ctx)
 	if syncSignal != nil {
-		go watch(ctx, syncPool, syncSignal)
+		go watch(ctx, syncPool, syncSignal, c)
 	}
 	return syncPool, nil
 }
 
-func watch(ctx context.Context, syncPool SyncPool, syncSignal <-chan struct{}) {
+func watch(ctx context.Context, syncPool SyncPool, syncSignal <-chan struct{}, c *healthChecker) {
 	for {
 		select {
 		case <-ctx.Done():
 			return
 		case <-syncSignal:
 			if err := syncPool.SyncOnce(ctx); err != nil {
```

Comment:
Is an error equal to not live? The question is, do we think restarting the pod (due to liveness probe failure) is the right thing to do if we the sync fails? Or do we think it's better for the pool to keep trying?

On a side note, it seems that the SyncOnce method never returns an error... The `errs` counter is always 0 I believe.

Reply:
We should write `SyncOnce` to be robust enough to not easily return error. As you noticed, we don't really return error today (let me also remove the useless `errs`). The error should only be returned on unrecoverable failures in which case we would want a pod restart. Make sense?

Reply:
If I'm interpreting the probe settings correctly, the container will only be restarted if this fails 3 times in a row.